### PR TITLE
benchmark against ajv

### PR DESCRIPTION
**Pull request by sakari**
# results

```
09:32 $ yarn jest test/benchmark-with-ajv/benchmark.spec.ts 
yarn run v1.22.19
$ /Users/sakari/work/oats/node_modules/.bin/jest test/benchmark-with-ajv/benchmark.spec.ts
  console.log
    # TaggedObjectUnion with props 100 levels 2

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:56:11)

  console.log
    ┌─────────┬─────────────────────┬──────────────┬───────────────────┬──────────┬─────────┐
    │ (index) │      Task Name      │   ops/sec    │ Average Time (ns) │  Margin  │ Samples │
    ├─────────┼─────────────────────┼──────────────┼───────────────────┼──────────┼─────────┤
    │    0    │ 'ajv nested types'  │ '10,491,435' │ 95.31583810441099 │ '±0.05%' │ 5245718 │
    │    1    │ 'oats nested types' │ '1,435,701'  │ 696.5238408713777 │ '±0.76%' │ 717851  │
    └─────────┴─────────────────────┴──────────────┴───────────────────┴──────────┴─────────┘

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:58:11)

  console.log
    # NestedObject with props 100 levels 2

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:56:11)

  console.log
    ┌─────────┬─────────────────────┬─────────┬───────────────────┬──────────┬─────────┐
    │ (index) │      Task Name      │ ops/sec │ Average Time (ns) │  Margin  │ Samples │
    ├─────────┼─────────────────────┼─────────┼───────────────────┼──────────┼─────────┤
    │    0    │ 'ajv nested types'  │  '516'  │ 1936670.36225897  │ '±0.93%' │   259   │
    │    1    │ 'oats nested types' │  '16'   │ 60503670.88317871 │ '±2.40%' │   10    │
    └─────────┴─────────────────────┴─────────┴───────────────────┴──────────┴─────────┘

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:58:11)

  console.log
    # TaggedObjectUnion with props 2 levels 5

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:56:11)

  console.log
    ┌─────────┬─────────────────────┬─────────────┬────────────────────┬──────────┬─────────┐
    │ (index) │      Task Name      │   ops/sec   │ Average Time (ns)  │  Margin  │ Samples │
    ├─────────┼─────────────────────┼─────────────┼────────────────────┼──────────┼─────────┤
    │    0    │ 'ajv nested types'  │ '9,815,546' │ 101.87919820852439 │ '±0.33%' │ 4907774 │
    │    1    │ 'oats nested types' │ '1,327,725' │ 753.1679111360003  │ '±1.59%' │ 663863  │
    └─────────┴─────────────────────┴─────────────┴────────────────────┴──────────┴─────────┘

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:58:11)

  console.log
    # NestedObject with props 2 levels 5

      at printTable (test/benchmark-with-ajv/benchmark.spec.ts:56:11)

  console.log
    ┌─────────┬─────────────────────┬───────────┬────────────────────┬──────────┬─────────┐
    │ (index) │      Task Name      │  ops/sec  │ Average Time (ns)  │  Margin  │ Samples │
    ├─────────┼─────────────────────┼───────────┼────────────────────┼──────────┼─────────┤
    │    0    │ 'ajv nested types'  │ '121,001' │  8264.36486024072  │ '±0.71%' │  60501  │
    │    1    │ 'oats nested types' │  '2,621'  │ 381422.08794798225 │ '±0.98%' │  1311   │
    └─────────┴─────────────────────┴───────────┴────────────────────┴──────────┴─────────┘
```

# notes

From the above oats seems to be within 10 to 100 times slower than AJV which is not unsurprising as AJV just validates whereas oats parses and constructs new object instances from the input. 

Not depicted above is the case where oats cannot figure out the tag for the tagged union but AJV can, eg when the tag property is not required, in that case oats can be some 1M times less performant -- oast does not really have good debugging output on whether it has found a tag for unions or not.

### CODE REVIEW

**Review by lautis — COMMENTED** [review]
Could running the benchmarks be a bit easier? Requiring erb sounds cumbersome evenif prevalent at Smartly.

File: test/benchmark-with-ajv/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Benchmark against ajv
+
+run `erb api.yml.erb > api.yml` in this directory
```
[the remark this answers is not in the document]
Comment (sakari) [reply test/benchmark-with-ajv/README.md:3]: I could try to figure something installable from node. Doing large template strings is not that much fun

[the remark this answers is not in the document]
Comment (sakari) [reply test/benchmark-with-ajv/README.md:3]: although running the erb template is only needed when changing the template as the output is committed to git

[the remark this answers is not in the document]
Comment (sakari) [reply test/benchmark-with-ajv/README.md]: changed to use ejs and generate the api.yml on generate

[the remark this answers is not in the document]
Comment (lautis) [reply test/benchmark-with-ajv/README.md]: If we don't care if the generated api.yml file is beautiful, wouldn't using JS + JSON.stringify do the job given that YAML is a superset of JSON?

**Review by lautis — APPROVED** [review]

**Pull request merged**